### DISCUSSION TITLE
test for quadratic convergence

### DISCUSSION THREAD

**Discussion by GhiathMonnet** [thelfer/tfel#382] [General]
Hi,
I have just completed the implementation of a visco-elastoplastic behavior.  
How can I check in a standard way the quadratic convergence of this law ?
Many thanks.   Ghiath

**Comment by thelfer** [comment]
Hi,

I assume that your are using a DSL from the `implicit` family. I also assume that you are taking about the convergence of the Newton algorithm of the behaviour, not the one of your solver. 

There are several methods:

1. You can compile the behaviour using the `--debug` flag as follows:
    ```
    $ mfront --obuild --debug --interface=....
    ```
    This will print the evolution of the residual during the resolution. With a bit of experience, you will be able to recognize a quadratic convergence
2. If you use the automatic computation of the consistent tangent operator, another good indicator is the decrease of the residual in the equilibrium iterations: the automatic computation of the consistent tangent operator relies on the Jacobian to compute the consistent tangent operator (see https://thelfer.github.io/tfel/web/implicit-dsl.html#sec:tangent_operator:small_strain_behaviours), so the consistent tangent operator is good only if Jacobian matrix is.
3. The better method is however to check the quality of the jacobian by a comparison with a numerical approximation, as follows:
    ```cxx
    mfront --obuild --@CompareToNumericalJacobian=true --@PerturbationValueForNumericalJacobianComputation =1.e-7--interface=
    ```
    See the online doc for details (example: `mfront --help-keyword=Implicit:@CompareToNumericalJacobian`)

**Reply by GhiathMonnet** [reply]
Thanks Thomas,
Sorry I forgot to specify that I am using @DSL IsotropicStrainHardeningMisesCreep
I am using mtest to check the convergence of the law. 
Cheers.
Ghiath

**Reply by thelfer** [reply]
Ok. So you mostly have two first options available.

For the second option, you can compare the value of the tangent operator to a numerical approximation in **`MTest`** as follows:

```
$ mtest --@CompareToNumericalTangentOperator=true --@NumericalTangentOperatorPerturbationValue=1.e-7 ...
```
